KVM: Add support for reporting CPU usage of domain

cshari-zededa opened the pull request:
Signed-off-by: Hariharasubramanian C S <cshari@zededa.com>

rvs commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -626,6 +626,31 @@ func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
 
 	for dom, pid := range ctx.domains {
+		//Fetch cpu usage from top output
+		//FIXME: Once we migrate to kvmtop util, remove this logic
```
Let me file a tracker story and assign it to you, ok @cshari-zededa ?

cshari-zededa replied:
> No, I don't want more screen scraping of some top command output. Far too brittle.
> gopsutil seems reasonable. Let's do that.

Done. and updated the PR.

rvs commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -626,6 +626,31 @@ func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
+		cmdOutputBytes, err := wrap.Command("/usr/bin/top", "-n", "1").CombinedOutput()
+		if err != nil {
+			return res, err
+		}
```
I'd rather not bail at this point, but fill the CPU portion with 0 as it used to be (to a point where it may be useful to flip the logic in the memory account as well -- make sure that if we have useful CPU statistics we report that even if gathering memory accounting fails)

cshari-zededa replied:
address this point.

rvs commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -626,6 +626,31 @@ func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
+			if percentageStr != nil {
+				pc, err := strconv.Atoi(strings.TrimSuffix(percentageStr[0], "%"))
+				if err != nil {
+					return res, err
```
ditto

cshari-zededa replied:
addressed.

eriknordmark commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -635,12 +660,18 @@ func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
 		if err != nil {
 			return res, logError("can't access stats for domain %s PID %d (%v)", dom, pid, err)
 		}
+		usedMemory := uint32(roundFromBytesToMbytes(uint64(s.ResidentMemory())))
+		availableMemory := uint32(roundFromBytesToMbytes(uint64(s.VirtualMemory())))
+		usedMemoryPerc := float64(0.0)
```
Does gopsutil's memory function return the same thing? I'm trying to have less variation in how we retrieve stats from Linux.

cshari-zededa replied:
moved to gopsutil.

eriknordmark commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -626,6 +626,31 @@ func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
 	for dom, pid := range ctx.domains {
+		//Fetch cpu usage from top output
+		//FIXME: Once we migrate to kvmtop util, remove this logic
+		cmdOutputBytes, err := wrap.Command("/usr/bin/top", "-n", "1").CombinedOutput()
+		if err != nil {
+			return res, err
+		}
```
Instead of screen-scraping top, can we use gopsutil to get the cpu usage for the pid?

cshari-zededa replied:
moved to gopsutil

rvs commented:
Hey @eriknordmark if we standardize on anything it should be kvmtop library, not gopsutils in my view -- but I agree standardizing is important and I think that's what @cshari-zededa will de doing long term.

rvs commented:
As per @eriknordmark comment today -- kvmtop is sadly licensed under GPL which means we will have to consider that if we want to link it as a library.

Thus, I suggest we go ahead with this PR (sans comments) and look into standardizing on a solution in the meantine.

Sounds good?

cshari-zededa commented:
> As per @eriknordmark comment today -- kvmtop is sadly licensed under GPL which means we will have to consider that if we want to link it as a library.
> 
> Thus, I suggest we go ahead with this PR (sans comments) and look into standardizing on a solution in the meantine.
> 
> Sounds good?

Hi Roman, 
Sure. Right now I addressing your comments, will update the PR shortly after fixing the places where we bail out due to error, to make sure we return available metrics at least. 
We can evaluate on standardization in the meantime.
Thanks
Hari

eriknordmark commented:
No, I don't want more screen scraping of some top command output. Far too brittle. 
gopsutil seems reasonable. Let's do that.

cshari-zededa commented:
@rvs @eriknordmark please check it out, I've updated the PR

eriknordmark commented on pkg/pillar/hypervisor/kvm.go:
```diff
@@ -616,31 +616,74 @@ func (ctx kvmContext) GetHostCPUMem() (types.HostMemory, error) {
-func (ctx kvmContext) GetDomsCPUMem() (map[string]types.DomainMetric, error) {
-	// for more precised measurements we should be using https://github.com/cha87de/kvmtop
-	res := map[string]types.DomainMetric{}
+//Fetch cpu usage from top output
```
Comment is out of date

rvs commented:
Fair enough @eriknordmark  -- I'll pull this in and fix the comment myself -- since apparently this could benefit Raghu and I need a build for him.
